docs/painless/painless-basic-syntax.asciidoc

```diff
@@ -0,0 +1,304 @@
+----
+
+*Examples:*
+
+String literals using both single-quotes and double-quotes.
+
+[source,Java]
+----
+"double-quoted String literal"
+'single-quoted String literal'
+"\"double-quoted String with escaped double-quotes\" and backslash: \\"
+'\'single-quoted String with escaped single-quotes\' and backslash \\'
+"double-quoted String with non-escaped 'single-quotes'"
+'single-quoted String with non-escaped "double-quotes"'
+----
+
+[[characters]]
+===== Characters
+
+Use the <<casting, casting operator>> to convert string literals or
+<<string-type, String type>> values into <<primitive-types, char type>> values.
+<<string-type, String type>> values converted into
+<<primitive-types, char type>> values must be exactly one character in length
+or an error will occur.
+
+*Examples:*
+
+Casting string literals into <<primitive-types, char type>> values.
+
+[source,Java]
+----
+(char)"C"
+(char)'c'
+----
+
+Casting a <<string-type, String type>> value into a
+<<primitive-types, char type>> value.
+
+[source,Java]
+----
+String s = "s";
+char c = (char)s;
+----
+
+[[variables]]
+==== Variables
+
+Variables in Painless must be declared and can be statically or <<dynamic-types,
+dynamically typed>>.
+
+[[variable-identifiers]]
+===== Variable Identifiers
+
+Specify variable identifiers using the following grammar. Variable identifiers
+must start with a letter or underscore. You cannot use <<keywords, keywords>> or
+<<painless-types, types>> as identifiers.
+
+*Grammar:*
+[source,ANTLR4]
+----
+ID: [_a-zA-Z] [_a-zA-Z-0-9]*;
+----
+
+*Examples:*
+[source,Java]
+----
+_
+a
+Z
+id
+list
+list0
+MAP25
+_map25
+----
+
+[[variable-declaration]]
+===== Variable Declaration
+
+Variables must be declared before you use them. The format is `type-name
+identifier-name`. To declare multiple variables of the same type, specify a
+comma-separated list of identifier names. You can immediately assign a value to
+a variable when you declare it.
+
+*Grammar:*
+[source,ANTLR4]
+----
+type: ID ('[' ']')*;
+declaration : type ID (',' ID)*;
+----
+
+*Examples:*
+[source,Java]
+----
+int x;        // Declare a variable with type int and id x
+List y;       // Declare a variable with type List and id y
+int x, y, z;  // Declare variables with type int and ids x, y, and z
+def[] d;      // Declare the variable d with type def[]
+int i = 10;   // Declare the int variable i and set it to the int literal 10
+----
+
+[[variable-assignment]]
+===== Variable Assignment
+
+Use the equals operator (`=`) to assign a value to a variable. The format is
+`identifier-name = value`. Any value expression can be assigned to any variable
+as long as the types match or the expression's type can be implicitly cast to
+the variable's type. An error occurs if the types do not match.
+
+*Grammar:*
+[source,ANTLR4]
+----
+assignment: ID '=' expression
+----
+
+
+*Examples:*
+
+Assigning a literal of the appropriate type directly to a declared variable.
```

Comment:
I'd probably replace the trailing `.` with a `:` because it is describing the block. I'm not sure what feels more right to be honest.

Reply:
I'm going to stick with standard sentence notation as it's more natural if there's more than one sentence.

Reply:
As long as it's consistent, either (or even none) works. I'd be inclined to drop the colons after the Grammar and Examples headings, they aren't really necessary.

Reply:
Will remove the colons from example and grammar as I work through editing the sections.